### PR TITLE
App/133 component tests

### LINKED ISSUE #133: app: Component Tests for Reusable UI Components
# Component Tests for Reusable UI Components

## Summary

Write component-level tests for the three reusable UI components in `frontend/components/`:
**Button**, **TextField**, and **Label**. These tests use `@testing-library/react-native` to
render each component in isolation and verify its behavior, prop handling, accessibility
attributes, and visual states.

## Motivation

`Button`, `TextField`, and `Label` are the building blocks used across every screen in the app
(login, registration, settings, etc.). If a prop is ignored, an accessibility attribute is
missing, or a state like `disabled` / `loading` doesn't work correctly, the bug will silently
propagate to every screen that uses the component. Component-level tests provide a safety net
that catches these regressions early.

### Current Component Inventory

| Component | Location | Props | Key Behaviors |
|-----------|----------|-------|---------------|
| **Button** | `components/Button/Button.tsx` | `onPress`, `title`, `variant`, `size`, `disabled`, `loading`, `accessibilityLabel`, `accessibilityHint`, `style`, `textStyle` | Press handling, disabled/loading states, variant styles (primary/secondary/danger), size styles (small/medium/large), accessibility |
| **TextField** | `components/TextField/TextField.tsx` | `value`, `onChangeText`, `placeholder`, `label`, `required`, `error`, `disabled`, `secureTextEntry`, `keyboardType`, `autoCapitalize`, `accessibilityLabel`, `accessibilityHint`, `rightIcon`, + more | Text input, focus/blur, error display with `accessibilityRole="alert"`, disabled state, secure entry toggle, label with asterisk |
| **Label** | `components/Label/Label.tsx` | `children`, `variant`, `color`, `style`, `accessibilityLabel` | Variant rendering (screenTitle, sectionHeader, formHeader, formLabel, error, caption, body), color override, accessibility |

### Why These Components First

- They are **isolated** — no context providers, no navigation, no API calls needed.
- They are **heavily reused** — every screen depends on at least one of them.
- They exercise **accessibility props** — verifying `accessibilityRole`, `accessibilityLabel`,
  `accessibilityState`, and `accessibilityHint` coverage.
- They establish **testing patterns** (rendering, querying, event firing) that will be reused
  in Issue 5 (integration tests).

## Dependencies

| Dependency | Required? | Notes |
|------------|-----------|-------|
| **Issue 1** | **Yes** | Jest + RNTL setup must be complete. |
| Issue 2 | No | Static analysis CI is independent. |
| Issue 3 | No | Service unit tests can be done in parallel. |

### `components/Button/__tests__/Button.test.tsx`

- [ ] **Rendering:**
  - Renders the `title` text inside the button.
  - Uses `accessibilityLabel` when provided; falls back to `title` when not.
- [ ] **Press behavior:**
  - Calls `onPress` when the button is pressed.
  - Does **not** call `onPress` when `disabled` is `true`.
  - Does **not** call `onPress` when `loading` is `true`.
- [ ] **Loading state:**
  - Shows an `ActivityIndicator` when `loading` is `true`.
  - Hides the title text visually (via `opacity: 0`) during loading.
  - Sets `accessibilityState.disabled` to `true` during loading.
- [ ] **Disabled state:**
  - Sets `accessibilityState.disabled` to `true`.
  - Applies `opacity: 0.5` container style.
- [ ] **Accessibility:**
  - `accessibilityRole` defaults to `"button"`.
  - `accessibilityHint` is forwarded when provided.
- [ ] **Variants:** Renders without error for each variant: `"primary"`, `"secondary"`,
  `"danger"`.
- [ ] **Sizes:** Renders without error for each size: `"small"`, `"medium"`, `"large"`.

### `components/TextField/__tests__/TextField.test.tsx`

- [ ] **Rendering:**
  - Renders `placeholder` text in the input.
  - Renders `label` text above the input when provided.
  - Renders required asterisk (`*`) next to the label when `required` is `true`.
  - Does not render asterisk when `required` is `false` (default).
- [ ] **Input behavior:**
  - Calls `onChangeText` when text is entered.
  - Forwards `secureTextEntry` prop to the underlying `TextInput`.
  - Forwards `keyboardType`, `autoCapitalize`, `autoComplete` props.
  - Calls `onSubmitEditing` when the return key is pressed.
- [ ] **Error state:**
  - Displays `error` text below the input when provided.
  - Error text has `accessibilityRole="alert"`.
  - Error text has `accessibilityLiveRegion="polite"`.
  - Error text has `accessibilityLabel` of `"Error: {message}"`.
  - Appends error to `accessibilityHint` on the input itself.
- [ ] **Disabled state:**
  - Input is not editable when `disabled` is `true`.
  - `accessibilityState.disabled` is `true`.
- [ ] **Accessibility fallbacks:**
  - `accessibilityLabel` falls back to `label` when `accessibilityLabel` prop is not set.
  - `accessibilityLabel` falls back to `placeholder` when neither prop nor `label` is set.
- [ ] **Right icon:**
  - Renders `rightIcon` element when provided.
  - Input has extra right padding when `rightIcon` is present.
- [ ] **Ref forwarding:**
  - `React.createRef()` can be passed and `.current` points to the `TextInput`.

### `components/Label/__tests__/Label.test.tsx`

- [ ] **Rendering:**
  - Renders `children` text content.
  - Renders React element children (nested `Text` components).
- [ ] **Variants:**
  - Renders without error for each variant: `"screenTitle"`, `"sectionHeader"`,
    `"formHeader"`, `"formLabel"`, `"error"`, `"caption"`, `"body"`.
  - Each variant applies its expected font size / weight (can snapshot or check style).
- [ ] **Color override:**
  - Applies `color` prop to the text style when provided.
  - Uses the variant's default color when `color` is not provided.
- [ ] **Custom style:**
  - Merges `style` prop with variant styles.
- [ ] **Accessibility:**
  - `accessibilityRole` is set to `"text"`.
  - `accessibilityLabel` is forwarded when provided.

### Example: `Button.test.tsx`

```tsx
import * as React from "react";
import { render, fireEvent, screen } from "@testing-library/react-native";
import Button from "../Button";

describe("Button", () => {
  const defaultProps = {
    onPress: jest.fn(),
    title: "Test Button",
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  // --- Rendering ---

  it("renders the title text", () => {
    render(<Button {...defaultProps} />);
    expect(screen.getByText("Test Button")).toBeTruthy();
  });

  it("uses custom accessibilityLabel when provided", () => {
    render(<Button {...defaultProps} accessibilityLabel="Custom Label" />);
    expect(screen.getByLabelText("Custom Label")).toBeTruthy();
  });

  it("falls back to title for accessibilityLabel", () => {
    render(<Button {...defaultProps} />);
    expect(screen.getByLabelText("Test Button")).toBeTruthy();
  });

  // --- Press behavior ---

  it("calls onPress when pressed", () => {
    render(<Button {...defaultProps} />);
    fireEvent.press(screen.getByRole("button", { name: "Test Button" }));
    expect(defaultProps.onPress).toHaveBeenCalledTimes(1);
  });

  it("does not call onPress when disabled", () => {
    render(<Button {...defaultProps} disabled />);
    fireEvent.press(screen.getByRole("button"));
    expect(defaultProps.onPress).not.toHaveBeenCalled();
  });

  it("does not call onPress when loading", () => {
    render(<Button {...defaultProps} loading />);
    fireEvent.press(screen.getByRole("button"));
    expect(defaultProps.onPress).not.toHaveBeenCalled();
  });

  // --- Accessibility ---

  it("has accessibilityRole button by default", () => {
    render(<Button {...defaultProps} />);
    expect(screen.getByRole("button")).toBeTruthy();
  });

  it("sets accessibilityState.disabled when disabled", () => {
    render(<Button {...defaultProps} disabled />);
    const button = screen.getByRole("button");
    expect(button.props.accessibilityState).toEqual(
      expect.objectContaining({ disabled: true })
    );
  });

  // --- Variants ---

  it.each(["primary", "secondary", "danger"] as const)(
    "renders %s variant without error",
    (variant) => {
      expect(() =>
        render(<Button {...defaultProps} variant={variant} />)
      ).not.toThrow();
    }
  );

  // --- Sizes ---

  it.each(["small", "medium", "large"] as const)(
    "renders %s size without error",
    (size) => {
      expect(() =>
        render(<Button {...defaultProps} size={size} />)
      ).not.toThrow();
    }
  );
});
```

### Example: `TextField.test.tsx` (partial)

```tsx
import * as React from "react";
import { render, fireEvent, screen } from "@testing-library/react-native";
import TextField from "../TextField";

describe("TextField", () => {
  const defaultProps = {
    value: "",
    onChangeText: jest.fn(),
  };

  it("renders placeholder text", () => {
    render(<TextField {...defaultProps} placeholder="Enter text" />);
    expect(screen.getByPlaceholderText("Enter text")).toBeTruthy();
  });

  it("displays error with alert role", () => {
    render(<TextField {...defaultProps} error="Required field" />);
    const errorText = screen.getByText("Required field");
    expect(errorText.props.accessibilityRole).toBe("alert");
  });

  it("shows required asterisk when required", () => {
    render(<TextField {...defaultProps} label="Email" required />);
    expect(screen.getByText("*")).toBeTruthy();
  });

  it("forwards ref to TextInput", () => {
    const ref = React.createRef<any>();
    render(<TextField {...defaultProps} ref={ref} placeholder="Test" />);
    expect(ref.current).toBeTruthy();
  });
});
```

### Testing Patterns & Conventions

**Query priority (from RNTL best practices):**

1. `getByRole` — best for accessibility; queries by `accessibilityRole`.
2. `getByLabelText` — queries by `accessibilityLabel`.
3. `getByText` — queries visible text content.
4. `getByPlaceholderText` — for text inputs.
5. `getByTestId` — last resort; avoid if possible.

**`React.createElement` note:**
The production components use `React.createElement(...)`. Test files can use JSX freely — RNTL
doesn't care how the element was created. If the TypeScript config doesn't support JSX in test
files, add `"jsx": "react-jsx"` to `compilerOptions` or create a `tsconfig.spec.json`.

### File Structure After This Issue

```
frontend/components/
├── Button/
│   ├── __tests__/
│   │   └── Button.test.tsx
│   ├── Button.tsx
│   ├── index.ts
│   └── styles.ts
├── TextField/
│   ├── __tests__/
│   │   └── TextField.test.tsx
│   ├── TextField.tsx
│   ├── index.ts
│   └── styles.ts
└── Label/
    ├── __tests__/
    │   └── Label.test.tsx
    ├── Label.tsx
    ├── index.ts
    └── styles.ts
```

### PR DESCRIPTION
**Pull request by rtipper1**
## Component Tests for Reusable UI Components

closes #133 

Adds component-level tests for reusable UI components using Jest and React Native Testing Library.

**Test Files Added:**
- `__tests__/Button.test.tsx` - tests for Button component
- `__tests__/TextField.test.tsx` - tests for TextField component
- `__tests__/Label.test.tsx` - tests for Label component

**Coverage:**
- **Button**: Rendering, press behavior, loading/disabled states, accessibility, variants (primary/secondary/danger), sizes (small/medium/large), custom styles
- **TextField**: Rendering, input behavior, error states, disabled states, accessibility fallbacks, right icon support, ref forwarding, focus/blur behavior
- **Label**: Rendering, all 7 variants (screenTitle, sectionHeader, formHeader, formLabel, error, caption, body), color override, custom styles, accessibility

**Results:** 91 tests total, all passing

All tests are located in `frontend/__tests__/` and verify component behavior, accessibility, and edge cases.

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -69,7 +69,7 @@
     "jest-expo": "^54.0.17",
     "jest-junit": "^16.0.0",
     "jest-transform-stub": "^2.0.0",
-    "react-test-renderer": "^19.2.4",
+    "react-test-renderer": "^19.1.0",
```
Comment (AkashK321) [inline frontend/package.json]: Just noting that this version was also changed in another PR, don't remember to what, but might be prudent to make sure the versions are consistent.

**Review by AkashK321 — APPROVED** [review]
